msal/application.py

```diff
@@ -304,7 +304,30 @@ def __init__(
 
             .. admonition:: Supporting raw assertion obtained from elsewhere
 
+                *Added in version 1.32.0*:
+
+                MSAL Python accepts a callback which shall obtain an assertion::
+
+                    import msal
+
+                    def assertion_callback():
```

Comment:
Can you add claims as input to this callback? This will help implementers of FIC perform token revocation, where the claims need to be take into account by the assertion provider as well.

Reply:
We can create an adjustable callback as a class, so that it can maintain its state. The crux is that the MSAL code path calling this callback does not know what proper values to feed in. So, the app developer - or a higher-level API - will need to expect an assertion error from eSTS, and then adjust that callback instance accordingly.

```python
class AdjustsableCallback:
    extra_claims = None

    def callback(self):
        return "A new assertion based on extra claims: " + self.extra_claims

ac = AdjustableCallback()
cca = ConfidentialClientApplication(..., client_credential={"client_assertion": ac.callback})

result = cca.acquire_token_for_client(...)
if result.get("error"):
    ac.extra_claims = get_new_claims_somehow()
    result = cca.acquire_token_for_client(...)
```

Do you want me to also document this advance scenario into the docs this time?

Reply:
The claims have to flow from MSAL's acquire_token_for_client call into the callback. An example would be good. 

Here's the equivalent .NET code


```csharp

// singleton
var cca = ConfidentialCLientApp(...)
                .WithClientAssertion(
                        (input) => {
                              string claims = input.Claims; 
                              GetAssertionFromMsiOrWhatever(claims);
                  });


// later in the code
var authResult = cca.acquireTokenForClient();  // ok, callback invoked, claims is empty

// call downstream API
// 401 + WWW-AUthenticate returned
// parse header and extract claims

// get new token with claims challenge
authResult = cca.acquireTokenForClient(claims); // ok, callback is invoked, claims is not empty
```